[SPARK-38336][SQL] Support DEFAULT column values in CREATE/REPLACE TABLE statements

dtenedor opened the pull request:
### What changes were proposed in this pull request?

Extend CREATE TABLE and REPLACE TABLE statements to support columns with DEFAULT values. This information will be stored in the column metadata.

### Why are the changes needed?

This builds the foundation for future work (not included in this PR) to support INSERT INTO statements, which may then omit the default values or refer to them explicitly with the DEFAULT keyword, in which case the Spark analyzer will automatically insert the appropriate corresponding values in the right places.

Example:
```
CREATE TABLE T(a INT DEFAULT 4, b INT NOT NULL DEFAULT 5);
INSERT INTO T VALUES (1);
INSERT INTO T VALUES (1, DEFAULT);
INSERT INTO T VALUES (DEFAULT, 6);
SELECT * FROM T;
(1, 5)
(1, 5)
(4, 6)
```

### How was this patch tested?

This change is covered by new and existing unit test coverage as well as new INSERT INTO query test cases covering a variety of positive and negative scenarios.

dtenedor committed:
.idea/vcs.xml

dtenedor committed:
Merge branch 'apache:master' into master

dtenedor committed:
Support DEFAULT column values in CREATE/REPLACE TABLE and INSERT INTO statements.

dtenedor committed:
Fix example in comment.

dtenedor committed:
update

dtenedor committed:
Restore project files.

dtenedor committed:
Restore project files.

dtenedor committed:
Restore project files.

dtenedor committed:
Restore project files.

dtenedor committed:
Restore project files.

AmplabJenkins commented:
Can one of the admins verify this patch?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala:
```diff
@@ -0,0 +1,266 @@
+   * @return the updated statement with DEFAULT column references replaced with their values.
+   */
+  def ReplaceExplicitDefaultColumnValues(insert: InsertIntoStatement,
+                                         catalog: SessionCatalog): InsertIntoStatement = {
```
nit: 4 spaces indent

dtenedor replied:
Done.

dtenedor committed:
Update to support implicit DEFAULT columns of NULL value.

dtenedor committed:
Update to support implicit DEFAULT columns of NULL value.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala:
```diff
@@ -0,0 +1,266 @@
+   * @param statementType which type of statement we are running, such as INSERT; useful for errors.
+   * @return Result of the analysis and constant-folding operation.
+   */
+  private def Analyze(colExpr: Expression, colText: String, colName: String,
```
Done.

dtenedor committed:
Update to support implicit DEFAULT columns of NULL value.

dtenedor committed:
Update to support implicit DEFAULT columns of NULL value.

dtenedor committed:
Update to support implicit DEFAULT columns of NULL value.

dtenedor committed:
Update to support implicit DEFAULT columns of NULL value.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala:
```diff
@@ -0,0 +1,287 @@
+   * @param statementType name of the statement being processed, such as INSERT; useful for errors.
+   * @return a copy of `tableSchema` with field metadata updated with the constant-folded values.
+   */
+  def ConstantFoldDefaultExpressions(tableSchema: StructType, statementType: String): StructType = {
```
All the method names should start with lower case words.
`ConstantFoldDefaultExpressions` => `constantFoldDefaultExpressions`

gengliangwang replied:
This method is actually duplicated with the optimizer rule `ConstantFolding`. What if we keep the foldable expression in metadata and leave it to `ConstantFolding`?

dtenedor replied:
My bad. Updated the method names. Thank you for your patience, I am new to the style guide :)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:
```diff
@@ -20,21 +20,31 @@ package org.apache.spark.sql.catalyst.analysis
       validateInputDimension(table)
       validateInputEvaluable(table)
       convert(table)
+    case i @ InsertIntoStatement(_, _, _, (_: UnresolvedInlineTable | _: Project), _, _) =>
```
Shall we only run the following when i.table is resolved? Thus we don't need to handle `UnresolvedInlineTable` in this rule.

dtenedor replied:
I found it was  necessary to run this when the inline table is unresolved since at this point, any "default" tokens comprise UnresolvedAttributes which are necessary to replace before we check analysis. I left a comment to this effect at this point in the code to help clarify.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:
```diff
@@ -20,21 +20,31 @@ package org.apache.spark.sql.catalyst.analysis
+      // [[ResolveInlineTables]] rule runs top-down. It is necessary to match against the entire
+      // INSERT INTO statement in this way in order to provide enough context to know what values
+      // to replace each explicit provided DEFAULT reference with.
+      DefaultColumns.ReplaceExplicitDefaultColumnValues(
```
How about: 
1. filling missing columns with "DEFAULT"
2. replace all the "DEFAULT" as the default column value

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala:
```diff
@@ -0,0 +1,287 @@
+      val result = folded match {
+        case Project(Seq(a: Alias), OneRowRelation()) => a.child
+      }
+      // Make sure the constant folding was successful.
```
Shall we also check the result data type is compatible with the column type? E.g. `i int default 'abc'`

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala:
```diff
@@ -0,0 +1,287 @@
+   * using the DEFAULT keyword as needed.
+   *
+   * Example:
+   * CREATE TABLE T(a INT DEFAULT 4, b INT NOT NULL DEFAULT 5);
```
Hi @srielau, I implemented it this way because the design doc appears to specify that the DEFAULT expression must be folded to a constant [1]. Is this intended? If not, that is OK, I can update this PR to skip the constant-folding and simply validate and store the original text in the column metadata instead.

[1] 
<img width="1316" alt="image" src="https://user-images.githubusercontent.com/99207096/158637309-38338720-2d13-4e32-9df1-8d59e2518f5a.png">

dtenedor replied:
Resolved this confusion offline with Serge. This should be evaluated at the INSERT INTO time:
```
Catalyst
The catalyst design is straight forward. The defaults are read in with the rest of the column info.
We leave DEFAULT keywords at DML as unresolved until analysis.
```
I can update the PR accordingly.

dtenedor replied:
This is done, the "current default" value obtained from the initial CREATE/REPLACE TABLE statement is now persisted directly without constant-folding. Instead, a new "exists default" column metadata now holds the constant-folded result separately. Following INSERT/UPDATE/MERGE statements can use the former while table scans can use the latter.

dtenedor committed:
Respond to review comments.

dtenedor committed:
Separate current-default from exists-default to separate entries in the column metadata.

dtenedor committed:
Add a SQLConf to control the behavior

dtenedor committed:
Move new SQLConf around to fix merge.

dtenedor committed:
Add some more test cases and separate configuration for injecting NULL values when the INSERT INTO statement provides fewer values than the table expects.

dtenedor committed:
Cleanup

dtenedor committed:
fix typo

dtenedor committed:
Fix typo

dtenedor committed:
Fix bug

dtenedor committed:
Fix bug

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Fix bugs

dtenedor committed:
Improve error messages + replace DEFAULTs in VALUES anywhere within the INSERT INTO query

dtenedor committed:
Merge branch 'apache:master' into master

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,354 @@
+    case i@InsertIntoStatement(_, _, _, _, _, _)
+      if SQLConf.get.enableDefaultColumns &&
+        i.query.collectFirst { case u: UnresolvedInlineTable => u }.isDefined =>
+      val newQuery = ResolveDefaultColumns(catalog, Some(i)).apply(i.query)
```
Note that `plan.resolveOperatorsWithPruning` will transform all the operators in `i` including query recursively.
Shall we move the methods `addMissingDefaultColumnValues` and `replaceExplicitDefaultColumnValues` to `object ResolveDefaultColumns` so that we don't need to create class `ResolveDefaultColumns` recursively.
The same applies to the following code.

dtenedor replied:
Good question: this is actually the intent to transform recursively, specifically only on the subplan in `i.query`. In this way, the following case matching against `UnresolvedInlineTable` may trigger anywhere that the corresponding `values` list appears in that subplan (e.g. `insert into t select * from values(...)` or more complex cases ending in `values`). I left a comment explaining this here. Otherwise in the more complex cases, the analyzer simply errors that `default` was not found in those cases.

If this SG to you, we can leave it as-is. Otherwise, if you prefer to only match the `i.query` against (`UnresolvedInlineTable` or `Project <- UnresolvedInlineTable`), let me know and I can make that change. We could possibly add another `case` then to `find` if any unresolved `default` references exist in any nested `values`, and return a more specific error message in that case to help guide the user.

gengliangwang commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala:
```diff
@@ -1025,16 +1024,29 @@ abstract class CatalogTestUtils {
   def newTable(name: String, db: String): CatalogTable = newTable(name, Some(db))
 
-  def newTable(name: String, database: Option[String] = None): CatalogTable = {
+  def newTable(name: String, database: Option[String] = None,
+               defaultColumns: Boolean = false): CatalogTable = {
```
```suggestion
  def newTable(
         name: String,
         database: Option[String] = None,
         defaultColumns: Boolean = false): CatalogTable = {
```

dtenedor replied:
Done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2781,13 +2782,18 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
       builder.putString("comment", _)
     }
+    // Add the 'DEFAULT expression' clause in the column definition, if any, to the column metadata.
+    Option(ctx.defaultExpression()).map(visitDefaultExpression).foreach { field =>
```
nit: We don't need function visitDefaultExpression. We can define a variable in the grammar file like
```
defaultExpression
    : DEFAULT defaultValue=expression
```
and then we can have 
```
Option(ctx.defaultExpression()).map(_.defaultValue.getText)
```
here

dtenedor replied:
Good idea, done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,354 @@
+  // This column metadata represents the default value for all existing rows in a table after a
+  // column has been added. This value is determined at time of CREATE TABLE, REPLACE TABLE, or
+  // ALTER TABLE ADD COLUMN, and never changes thereafter. The intent is for this "exist default"
+  // to be used by any scan when the columns in the source row are incomplete.
```
I don't quite understand this. What do you mean by "columns in the source row are incomplete"?

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2709,6 +2709,28 @@ object SQLConf {
     .createWithDefault(false)
 
+  val ENABLE_DEFAULT_COLUMNS =
+    buildConf("spark.sql.parser.enableDefaultColumns")
```
Sure, that SG. Done.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2709,6 +2709,28 @@ object SQLConf {
+        "update default values for specific columns. Following INSERT, MERGE, and UPDATE " +
+        "statements may then omit these values and their values will be injected automatically " +
+        "instead.")
+      .version("3.3.0")
```
Done (and for the USE_NULLS_... config below as well.)

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2709,6 +2709,28 @@ object SQLConf {
+      .createWithDefault(true)
+
+  val USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES =
+    buildConf("spark.sql.parser.useNullsForMissingDefaultColumnValues")
```
SG, Done.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2709,6 +2709,28 @@ object SQLConf {
+      .createWithDefault(true)
+
+  val USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES =
+    buildConf("spark.sql.parser.useNullsForMissingDefaultColumnValues")
```
SG, Done.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2709,6 +2709,28 @@ object SQLConf {
+        "columns in the target table, and the remaining columns will receive NULL values.")
+      .version("3.3.0")
+      .booleanConf
+      .createWithDefault(false)
```
Yeah this PR is already big :) we can leave this disabled by default for now, and then consider next whether we want to enable it. Without enabling this by default, this new functionality in this PR does not change the results of any INSERT INTO statements where the corresponding CREATE/REPLACE TABLE commands lack any DEFAULT values (which should be all the cases up until now).

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,354 @@
+ * @param insert the enclosing INSERT statement for which this rule is processing the query, if any.
+ */
+case class ResolveDefaultColumns(
+    catalog: SessionCatalog, insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
```
Done.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,354 @@
+ * @param insert the enclosing INSERT statement for which this rule is processing the query, if any.
+ */
+case class ResolveDefaultColumns(
+    catalog: SessionCatalog, insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
```
Done.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,354 @@
+ */
+case class ResolveDefaultColumns(
+    catalog: SessionCatalog, insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
```
SG, that makes the code simpler.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,354 @@
+ */
+case class ResolveDefaultColumns(
+    catalog: SessionCatalog, insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
```
SG, that makes the code simpler.

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2709,6 +2709,28 @@ object SQLConf {
+        "update default values for specific columns. Following INSERT, MERGE, and UPDATE " +
+        "statements may then omit these values and their values will be injected automatically " +
+        "instead.")
+      .version("3.3.0")
```
Done.

dtenedor committed:
Respond to code review comments.

Respond to code review comments.

Respond to code review comments.

Respond to code review comments.

Respond to code review comments.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,371 @@
+  }
+
+  // Helper method to check that an inline table has any explicit DEFAULT column references.
+  private def hasExplicitDefaultReferences(table: UnresolvedInlineTable): Boolean =
```
This method is not used.

dtenedor replied:
Good catch, removed this!

gengliangwang commented:
@dtenedor this PR looks good overall. 
However, it is quite big and it takes effort to review it carefully.
Do you mind splitting it into two PRs? One for creating tables and another one for inserting into.
I am pretty sure the create table one can be merged very soon.

dtenedor committed:
Remove unused method

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,366 @@
+  }
+
+  // This is a helper for the addMissingDefaultColumnValues methods above.
+  private def getDefaultExprs(numQueryOutputs: Int, schema: StructType,
```
param schema is not used

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:
```diff
@@ -846,6 +854,238 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
+      sql("insert into t values(1)")
+      checkAnswer(sql("select s + x from t where i = 1"), Seq(85L).map(i => Row(i)))
+    }
+    // The table has a partitioning column and a default value is injected.
```
Good question. It looks like the grammar does not support this. I added a test case and initially got a weird parsing error message. I added a rule in the grammar to catch this and return a better error message, and updated the new test case accordingly.

dtenedor committed:
Remove unused method

dtenedor committed:
Merge branch 'apache:master' into master

dtenedor committed:
Merge branch 'master' into default-cols-create-table

merge latest updates from master

dtenedor committed:
Remove changes related to INSERT INTO statements, those will take place in the next PR

dtenedor commented:
The work is now split into different PRs:

1. This PR contains only DEFAULT column support for CREATE/REPLACE TABLE statements: https://github.com/apache/spark/pull/35855
2. Separate PR supports DEFAULT columns for INSERT INTO statements: https://github.com/dtenedor/spark/pull/1
3. Third PR adds a query execution class to be used for injecting DEFAULT values during data source scans: https://github.com/apache/spark/pull/35928

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+  val EXISTS_DEFAULT_COLUMN_METADATA_KEY = "EXISTS_DEFAULT"
+  // Name of attributes representing explicit references to the value stored in the above
+  // CURRENT_DEFAULT_COLUMN_METADATA.
+  val CURRENT_DEFAULT_COLUMN_NAME = "DEFAULT"
```
nit: this is not used

dtenedor replied:
Done. (My bad, this was leftover from the child PR)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+   * @return a copy of `tableSchema` with field metadata updated with the constant-folded values.
+   */
+  def constantFoldCurrentDefaultsToExistDefaults(
+      tableSchema: StructType, statementType: String): StructType = {
```
```suggestion
      tableSchema: StructType,
      statementType: String): StructType = {
```

dtenedor replied:
Done.

gengliangwang commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala:
```diff
@@ -1025,16 +1024,30 @@ abstract class CatalogTestUtils {
   def newTable(name: String, db: String): CatalogTable = newTable(name, Some(db))
 
-  def newTable(name: String, database: Option[String] = None): CatalogTable = {
+  def newTable(name: String,
+               database: Option[String] = None,
+               defaultColumns: Boolean = false): CatalogTable = {
```
```suggestion
  def newTable(
      name: String,
      database: Option[String] = None,
      defaultColumns: Boolean = false): CatalogTable = {
```

dtenedor replied:
Done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+   * @param statementType name of the statement being processed, such as INSERT; useful for errors.
+   * @return a copy of `tableSchema` with field metadata updated with the constant-folded values.
+   */
+  def constantFoldCurrentDefaultsToExistDefaults(
```
This method is actually called from `DataSourceV2Strategy`.
Shall we move the whole object under `org.apache.spark.sql.catalyst.util` or some where else, instead of `org.apache.spark.sql.catalyst.analysis`?

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:
```diff
@@ -187,12 +190,14 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
 
     case ReplaceTable(ResolvedDBObjectName(catalog, ident), schema, parts, tableSpec, orCreate) =>
+      val newSchema: StructType =
+        ResolveDefaultColumns.constantFoldCurrentDefaultsToExistDefaults(schema, "CREATE TABLE")
```
I suggest passing the analyzer via function parameter
```
session.sessionState.analyzer
```
instead of creating a new one in object `ResolveDefaultColumns`
After supporting multiple catalogs, different catalogs will have different SQL functions which can be used in default values.

dtenedor replied:
Nice, this is more consistent. Thank you for suggesting this.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+   * @return Result of the analysis and constant-folding operation.
+   */
+  private def analyze(
+      field: StructField, statementType: String, foldConstants: Boolean = false): Expression = {
```
Also, the parameter `foldConstants` doesn't seem needed in this PR.

dtenedor replied:
Done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+    val colText: String = if (field.metadata.contains(CURRENT_DEFAULT_COLUMN_METADATA_KEY)) {
+      field.metadata.getString(CURRENT_DEFAULT_COLUMN_METADATA_KEY)
+    } else {
+      "NULL"
```
When will this happen? `EXISTS_DEFAULT_COLUMN_METADATA_KEY` exists and CURRENT_DEFAULT_COLUMN_METADATA_KEY doesn't exist.

dtenedor replied:
Good point. This should really just assume that `CURRENT_DEFAULT_COLUMN_METADATA_KEY` is populated. I updated the call site to check that explicitly and then deduplicated the two extractions of the column metadata here in this `analyze` method into one, which can skip the presence check and just extract the metadata value directly now.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+            s"${field.name} has a DEFAULT value of $colText which fails to resolve as a valid " +
+            s"expression: ${ex.getMessage}")
+    }
+    val analyzed = plan match {
```
How about using 
```
plan collectFirst
```

dtenedor replied:
Nice, that's more concise

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+    } catch {
+      case ex: ParseException =>
+        throw new AnalysisException(
+          s"Failed to execute $statementType command because the destination table column " +
```
Let's add an invalid test case for this

dtenedor replied:
Good idea, done. Also the other error cases.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+      if (foldConstants) ConstantFolding(analyzed) else analyzed
+    } catch {
+      case ex @ (_: ParseException | _: AnalysisException) =>
+        val colText: String = if (field.metadata.contains(CURRENT_DEFAULT_COLUMN_METADATA_KEY)) {
```
Let's add a test case for this

dtenedor replied:
Done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,154 @@
+      analyzer.checkAnalysis(analyzed)
+      if (foldConstants) ConstantFolding(analyzed) else analyzed
+    } catch {
+      case ex @ (_: ParseException | _: AnalysisException) =>
```
Why there is ParseException here

dtenedor replied:
This was leftover from before, not needed now. Removed.

dtenedor committed:
Respond to code review comments

dtenedor committed:
Add expected parse results in DDLParserSuite testing

dtenedor committed:
all

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2805,6 +2805,17 @@ object SQLConf {
     .booleanConf
     .createWithDefault(sys.env.get("SPARK_ANSI_SQL_MODE").contains("true"))
 
+  val ENABLE_DEFAULT_COLUMNS =
```
We need a test case with this conf as `false`, too

dtenedor replied:
Sure, this is done.

dtenedor committed:
add new test case

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,143 @@
+   * This method constant-folds the "current default" value, stored in the CURRENT_DEFAULT metadata
+   * of the "b" column, to "10", storing the result in the "exists default" value within the
+   * EXISTS_DEFAULT metadata of that same column. Meanwhile the "current default" metadata of this
+   * "b" column retains its original value of "5 + 5".
```
Why don't we do constant fold the "current default" as well?

dtenedor replied:
Good question.

The reason for constant-folding the EXISTS_DEFAULT is to make the end-user visible behavior the same after executing an ALTER TABLE ADD COLUMNS command with DEFAULT value as if the system had performed an exhaustive backfill of the provided value to all previously existing rows in the table instead. We choose to avoid doing that because it is a time-consuming and costly operation, electing instead to store the EXISTS_DEFAULT in the column metadata for future reference when querying data out of the data source. Each data source takes responsibility to provide the constant-folded value in the EXISTS_DEFAULT metadata for such columns where the value is not present in storage.

I have added this explanation to the comment here.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala:
```diff
@@ -0,0 +1,143 @@
+   * EXISTS_DEFAULT metadata of that same column. Meanwhile the "current default" metadata of this
+   * "b" column retains its original value of "5 + 5".
+   *
+   * @param tableSchema represents the names and types of the columns of the statement to process.
```
nit: let's have a comment on param `analyzer`

dtenedor replied:
Done.

gengliangwang commented:
LGTM overall. This is great work!

dtenedor committed:
add more comments

dtenedor commented:
Thanks for the review! Looks like this is ready to merge now at your convenience

gengliangwang commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala:
```diff
@@ -120,6 +121,67 @@ abstract class SessionCatalogSuite extends AnalysisTest with Eventually {
+
+      // Make sure that constant-folding default values does not take place when the feature is
+      // disabled.
+      withSQLConf(SQLConf.ENABLE_DEFAULT_COLUMNS.key -> "false") {
```
nit: actually I was expecting a test case which captures parser exception when the conf is disabled.

dtenedor replied:
SG, I added that one in the DDLParserSuite as well.

dtenedor committed:
Add new parser test to exercise when the feature is disabled

dtenedor committed:
Fix whitespace

gengliangwang reviewed (approved):
LGTM

gengliangwang commented:
Thanks, merging to master
